warehouse/query-core/src/main/java/datawave/query/jexl/functions/GeoWaveFunctionsDescriptor.java

```diff
@@ -219,4 +238,61 @@ public JexlNode apply(ByteArrayRange input) {
         
     }
     
+    protected static List<Geometry> getAllEnvelopeGeometries(Geometry geom) {
+        List<Geometry> geometries = new ArrayList<Geometry>();
+        if (geom.getNumGeometries() > 1) {
+            for (int geoIdx = 0; geoIdx < geom.getNumGeometries(); geoIdx++)
+                geometries.addAll(getAllEnvelopeGeometries(geom.getGeometryN(geoIdx)));
+        } else {
+            geometries.add(geom.getEnvelope());
+        }
+        return geometries;
+    }
+    
+    protected static List<List<Geometry>> combineOverlappedGeometries(List<Geometry> geometries) {
+        List<List<Geometry>> olpGeoms = new ArrayList<>();
+        while (!geometries.isEmpty()) {
+            // remove the first geometry
+            List<Geometry> overlapped = new ArrayList<>();
+            overlapped.add(geometries.remove(0));
+            
+            // find all geometries that intersect with this geometry, and as this combined set grows, we will need to
+            // reevaluate all the remaining geometries until no intersections are found, or until we run out of
+            // geometries
+            boolean geomAdded = true;
+            while (geomAdded) {
+                geomAdded = false;
+                // for each of the remaining geometries
+                for (Geometry srcGeom : geometries) {
+                    // check to see if this srcGeom intersects with any of the overlapped geometries
+                    for (Geometry ovlp : overlapped) {
```

Comment:
Some of this work is redundant.  Once an item is added to the overlapped set, it will be checked against all of the geometries on every subsequent pass.  Perhaps we can do some bookkeeping to avoid the extract intersection checks.

Reply:
I just pushed an updated implementation which should be better.  Let me know what you think.

Reply:
@ivakegg This should be much better now with my second revision.